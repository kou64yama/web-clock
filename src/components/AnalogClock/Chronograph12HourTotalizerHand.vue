<template>
  <g :transform="`rotate(${theta}, 50, 70)`">
    <path
      :d="`M 50 70 L 50 59`"
      stroke="#cccccc"
      stroke-width="0.7"
      stroke-linecap="round"
    />
    <circle cx="50" cy="70" r="1" fill="#cccccc" />
  </g>
</template>

<script lang="ts">
import { defineComponent, computed } from 'vue';
import { gear, HOUR } from './utils';

const reducer = gear(12 * HOUR, 360);

export default defineComponent({
  props: {
    duration: { type: Number, default: 0 },
  },
  setup: (props) => ({
    theta: computed(() => reducer(props.duration)),
  }),
});
</script>
