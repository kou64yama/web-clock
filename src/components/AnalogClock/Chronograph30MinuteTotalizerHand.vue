<template>
  <g :transform="`rotate(${theta}, 70, 50)`">
    <path
      :d="`M 70 50 L 70 39`"
      stroke="#cccccc"
      stroke-width="0.7"
      stroke-linecap="round"
    />
    <circle cx="70" cy="50" r="1" fill="#cccccc" />
  </g>
</template>

<script lang="ts">
import { defineComponent, computed } from 'vue';
import { gear, MINUTE } from './utils';

const reducer = gear(30 * MINUTE, 360);

export default defineComponent({
  props: {
    duration: { type: Number, default: 0 },
  },
  setup: (props) => ({
    theta: computed(() => reducer(props.duration)),
  }),
});
</script>
