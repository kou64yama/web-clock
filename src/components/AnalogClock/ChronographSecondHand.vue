<template>
  <g :transform="`rotate(${theta}, 50, 50)`">
    <path
      :d="`M 50 50 L 50 7`"
      stroke="#ff0000"
      stroke-width="0.5"
      stroke-linecap="round"
    />
    <circle cx="50" cy="50" r="1" fill="#ff0000" />
  </g>
</template>

<script lang="ts">
import { defineComponent, computed } from 'vue';
import { gear, MINUTE } from './utils';

const reducer = gear(MINUTE, 360);

export default defineComponent({
  props: {
    duration: { type: Number, default: 0 },
  },
  setup: (props) => ({
    theta: computed(() => reducer(props.duration)),
  }),
});
</script>
