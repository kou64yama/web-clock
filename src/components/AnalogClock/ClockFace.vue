<template>
  <g font-family="DM Serif Display">
    <circle cx="50" cy="50" r="50" fill="#222222" />
    <circle cx="50" cy="50" r="45" fill="#111111" />
    <circle cx="50" cy="70" r="13" fill="#222222" />
    <circle cx="30" cy="50" r="13" fill="#222222" />
    <circle cx="70" cy="50" r="13" fill="#222222" />
    <template v-for="i in range(0, 299)">
      <path
        v-if="i % 75 === 0"
        :key="`main.scale.${i}`"
        :d="`M 50 6 L 50 9`"
        stroke="#ff0000"
        stroke-width="0.7"
        :transform="`rotate(${(i * 360) / 300}, 50, 50)`"
      />
      <path
        v-else-if="i % 25 === 0"
        :key="`main.scale.${i}`"
        :d="`M 50 6 L 50 9`"
        stroke="#cccccc"
        stroke-width="0.5"
        :transform="`rotate(${(i * 360) / 300}, 50, 50)`"
      />
      <path
        v-else-if="i % 5 === 0"
        :key="`main.scale.${i}`"
        :d="`M 50 6 L 50 9`"
        stroke="#cccccc"
        stroke-width="0.2"
        :transform="`rotate(${(i * 360) / 300}, 50, 50)`"
      />
      <path
        v-else
        :key="`main.scale.${i}`"
        :d="`M 50 7 L 50 8`"
        stroke="#cccccc"
        stroke-width="0.2"
        :transform="`rotate(${(i * 360) / 300}, 50, 50)`"
      />
    </template>
    <template v-for="i in range(0, 11)" :key="`main.scale-symbol.${i}`">
      <path
        :d="`M 50 10 L 50 15`"
        stroke="#cccccc"
        stroke-width="2"
        :transform="`rotate(${i * 30}, 50, 50)`"
      />
    </template>
    <template v-for="i in range(0, 59)">
      <path
        v-if="i % 5 === 0"
        :key="`small-second.scale.${i}`"
        :d="`M 30 37.5 L 30 40`"
        stroke="#cccccc"
        stroke-width="0.3"
        :transform="`rotate(${i * 6}, 30, 50)`"
      />
      <path
        v-else
        :key="`small-second.scale.${i}`"
        :d="`M 30 37.5 L 30 39`"
        stroke="#cccccc"
        stroke-width="0.2"
        :transform="`rotate(${i * 6}, 30, 50)`"
      />
    </template>
    <template v-for="i in range(0, 29)">
      <path
        v-if="i % 5 === 0"
        :key="`30-minute-totalizer.scale.${i}`"
        :d="`M 70 37.5 L 70 40`"
        stroke="#cccccc"
        stroke-width="0.3"
        :transform="`rotate(${i * 12}, 70, 50)`"
      />
      <path
        v-else
        :key="`30-minute-totalizer.scale.${i}`"
        :d="`M 70 37.5 L 70 39`"
        stroke="#cccccc"
        stroke-width="0.2"
        :transform="`rotate(${i * 12}, 70, 50)`"
      />
    </template>
    <template v-for="i in range(0, 11)" :key="`12-hour-totalizer.scale.${i}`">
      <path
        :d="`M 50 57.5 L 50 59`"
        stroke="#cccccc"
        stroke-width="0.2"
        :transform="`rotate(${i * 30}, 50, 70)`"
      />
    </template>
    <template v-for="i in [30, 60]" :key="`small-second.label.${i}`">
      <text
        :x="30 + 7 * sin((PI * i) / 30)"
        :y="50 - 7 * cos((PI * i) / 30)"
        font-size="4"
        text-anchor="middle"
        dominant-baseline="central"
        fill="#cccccc"
        v-text="i"
      />
    </template>
    <template v-for="i in [10, 20, 30]" :key="`30-minute-totalizer.label.${i}`">
      <text
        :x="70 + 7 * sin((PI * i) / 15)"
        :y="50 - 7 * cos((PI * i) / 15)"
        font-size="4"
        text-anchor="middle"
        dominant-baseline="central"
        fill="#cccccc"
        v-text="i"
      />
    </template>
    <template v-for="i in [3, 6, 9, 12]" :key="`12-hour-totalizer.label.${i}`">
      <text
        :x="50 + 9 * sin((PI * i) / 6)"
        :y="70 - 9 * cos((PI * i) / 6)"
        font-size="4"
        text-anchor="middle"
        dominant-baseline="central"
        fill="#cccccc"
        v-text="i"
      />
    </template>
  </g>
</template>

<script lang="ts">
import { defineComponent } from 'vue';

const PI = Math.PI;
const range = (start: number, end: number) =>
  new Array(end - start + 1)
    .fill(0)
    .reduce<number[]>((acc) => [...acc, acc.length], []);
const sin = (x: number): number => Math.sin(x);
const cos = (x: number): number => Math.cos(x);

export default defineComponent({
  setup: () => ({ PI, range, sin, cos }),
});
</script>
