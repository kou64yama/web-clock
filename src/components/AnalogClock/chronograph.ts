import { computed, ComputedRef } from 'vue';
import { useLocalStorage } from '../../compositions/local-storage';

const STARTED = 'chronograph.started';
const DURATION = 'chronograph.duration';

interface Chronograph {
  duration: ComputedRef<number>;
  paused: ComputedRef<boolean>;
  startOrStop: () => void;
  reset: () => void;
}

export const useChronograph = (quartz: ComputedRef<number>): Chronograph => {
  const [rawStarted, setRawStarted] = useLocalStorage(STARTED);
  const [rawStored, setRawStored] = useLocalStorage(DURATION);

  const started = computed(() => {
    if (rawStarted.value === null) return null;
    const value = parseInt(rawStarted.value, 10);
    return Number.isFinite(value) ? value : null;
  });
  const stored = computed(() => {
    if (rawStored.value === null) return 0;
    const value = parseInt(rawStored.value, 10);
    return Number.isFinite(value) ? value : 0;
  });
  const duration = computed(() => {
    if (started.value === null) return stored.value;
    return stored.value + quartz.value - started.value;
  });
  const paused = computed(() => started.value === null);

  const start = () => {
    setRawStarted(`${quartz.value}`);
  };

  const stop = () => {
    if (started.value === null) return;
    setRawStored(`${stored.value + quartz.value - started.value}`);
    setRawStarted(null);
  };

  const reset = () => setRawStored(null);

  const startOrStop = () => (paused.value ? start() : stop());

  return { duration, paused, startOrStop, reset };
};
