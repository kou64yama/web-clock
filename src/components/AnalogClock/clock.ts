import {
  DeepReadonly,
  onBeforeUnmount,
  onMounted,
  readonly,
  Ref,
  ref,
} from 'vue';

interface Clock {
  quartz: DeepReadonly<Ref<number>>;
}

export const useClock = (): Clock => {
  let frame: number | null = null;
  const quartz = ref(0);

  const tick = () => {
    const value = Date.now();
    if (quartz.value !== value) {
      quartz.value = value;
    }

    frame = requestAnimationFrame(tick);
  };

  onMounted(tick);

  onBeforeUnmount(() => {
    if (frame !== null) cancelAnimationFrame(frame);
  });

  return { quartz: readonly(quartz) };
};
