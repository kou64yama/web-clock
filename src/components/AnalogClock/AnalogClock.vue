<template>
  <div>
    <svg
      xmlns="http://www.w3.org/2000/svg"
      :width="size"
      :height="size"
      :viewBox="`0 0 ${size} ${size}`"
      role="presentation"
    >
      <g :transform="`scale(${size / 100})`">
        <clock-face :size="size" />
        <short-hand :time="time" :size="size" />
        <long-hand :time="time" :size="size" />
        <second-hand :time="time" :size="size" />
      </g>
    </svg>
  </div>
</template>

<script lang="ts">
import { defineComponent } from "@vue/composition-api";
import ClockFace from "./ClockFace.vue";
import ShortHand from "./ShortHand.vue";
import LongHand from "./LongHand.vue";
import SecondHand from "./SecondHand.vue";

export default defineComponent({
  props: {
    time: { type: Number, default: 0 },
    size: { type: Number, default: 240 }
  },
  components: {
    ClockFace,
    ShortHand,
    LongHand,
    SecondHand
  }
});
</script>
