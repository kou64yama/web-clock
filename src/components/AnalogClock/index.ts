import AnalogClock from "./AnalogClock.vue";

export default AnalogClock;
export { useClock } from "./clock";
