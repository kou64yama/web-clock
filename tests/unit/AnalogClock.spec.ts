import { mount, createLocalVue } from "@vue/test-utils";
import CompositionApi from "@vue/composition-api";
import AnalogClock from "@/components/AnalogClock/index";

const localVue = createLocalVue();
localVue.use(CompositionApi);

describe("AnalogClock", () => {
  it("renders correctly", () => {
    const wrapper = mount(AnalogClock, {
      localVue,
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
