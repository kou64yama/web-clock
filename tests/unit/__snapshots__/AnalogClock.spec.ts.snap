// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnalogClock renders correctly 1`] = `
<div>
  <svg
    height="240"
    role="presentation"
    viewBox="0 0 240 240"
    width="240"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      transform="scale(2.4)"
    >
      <g
        size="240"
      >
        <path
          d="M 50 5 L 50 7"
          stroke="#000000"
          stroke-width="2"
          transform="rotate(0, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(6, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(12, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(18, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(24, 50, 50)"
        />
        <path
          d="M 50 5 L 50 7"
          stroke="#000000"
          stroke-width="2"
          transform="rotate(30, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(36, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(42, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(48, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(54, 50, 50)"
        />
        <path
          d="M 50 5 L 50 7"
          stroke="#000000"
          stroke-width="2"
          transform="rotate(60, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(66, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(72, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(78, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(84, 50, 50)"
        />
        <path
          d="M 50 5 L 50 7"
          stroke="#000000"
          stroke-width="2"
          transform="rotate(90, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(96, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(102, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(108, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(114, 50, 50)"
        />
        <path
          d="M 50 5 L 50 7"
          stroke="#000000"
          stroke-width="2"
          transform="rotate(120, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(126, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(132, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(138, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(144, 50, 50)"
        />
        <path
          d="M 50 5 L 50 7"
          stroke="#000000"
          stroke-width="2"
          transform="rotate(150, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(156, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(162, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(168, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(174, 50, 50)"
        />
        <path
          d="M 50 5 L 50 7"
          stroke="#000000"
          stroke-width="2"
          transform="rotate(180, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(186, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(192, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(198, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(204, 50, 50)"
        />
        <path
          d="M 50 5 L 50 7"
          stroke="#000000"
          stroke-width="2"
          transform="rotate(210, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(216, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(222, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(228, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(234, 50, 50)"
        />
        <path
          d="M 50 5 L 50 7"
          stroke="#000000"
          stroke-width="2"
          transform="rotate(240, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(246, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(252, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(258, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(264, 50, 50)"
        />
        <path
          d="M 50 5 L 50 7"
          stroke="#000000"
          stroke-width="2"
          transform="rotate(270, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(276, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(282, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(288, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(294, 50, 50)"
        />
        <path
          d="M 50 5 L 50 7"
          stroke="#000000"
          stroke-width="2"
          transform="rotate(300, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(306, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(312, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(318, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(324, 50, 50)"
        />
        <path
          d="M 50 5 L 50 7"
          stroke="#000000"
          stroke-width="2"
          transform="rotate(330, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(336, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(342, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(348, 50, 50)"
        />
        <path
          d="M 50 5 L 50 6"
          stroke="#000000"
          stroke-width="1"
          transform="rotate(354, 50, 50)"
        />
      </g>
       
      <g
        size="240"
        transform="rotate(0, 50, 50)"
      >
        <path
          d="M 50 50 L 50 30"
          stroke="#000000"
          stroke-width="2.5"
        />
      </g>
       
      <g
        size="240"
        transform="rotate(0, 50, 50)"
      >
        <path
          d="M 50 50 L 50 10"
          stroke="#000000"
          stroke-width="1.25"
        />
      </g>
       
      <g
        size="240"
        transform="rotate(0, 50, 50)"
      >
        <path
          d="M 50 50 L 50 8"
          stroke="#ff0000"
          stroke-width="0.5"
        />
         
        <circle
          cx="50"
          cy="50"
          fill="#ff0000"
          r="2"
        />
      </g>
    </g>
  </svg>
</div>
`;
